pkg/prometheus/statefulset.go

```diff
@@ -96,27 +96,41 @@ func prometheusNameByShard(name string, shard int32) string {
 func makeStatefulSet(
 	logger log.Logger,
 	name string,
-	p monitoringv1.Prometheus,
+	baseImage, tag, sha string,
+	retention monitoringv1.Duration,
+	retentionSize monitoringv1.ByteSize,
+	rules monitoringv1.Rules,
+	query *monitoringv1.QuerySpec,
+	allowOverlappingBlocks bool,
+	queryLogFile string,
+	thanos *monitoringv1.ThanosSpec,
+	disableCompaction bool,
+	p monitoringv1.PrometheusInterface,
 	config *operator.Config,
 	cg *ConfigGenerator,
 	ruleConfigMapNames []string,
 	inputHash string,
 	shard int32,
 	tlsAssetSecrets []string,
 ) (*appsv1.StatefulSet, error) {
-	if p.Spec.PortName == "" {
-		p.Spec.PortName = defaultPortName
+	cpf := p.GetCommonPrometheusFields()
+	objMeta := p.GetObjectMeta()
+	typeMeta := p.GetTypeMeta()
+
+	if cpf.PortName == "" {
+		cpf.PortName = defaultPortName
 	}
 
-	if p.Spec.Replicas == nil {
-		p.Spec.Replicas = &minReplicas
+	if cpf.Replicas == nil {
+		cpf.Replicas = &minReplicas
 	}
 	intZero := int32(0)
-	if p.Spec.Replicas != nil && *p.Spec.Replicas < 0 {
-		p.Spec.Replicas = &intZero
+	if cpf.Replicas != nil && *cpf.Replicas < 0 {
+		cpf.Replicas = &intZero
 	}
 
-	spec, err := makeStatefulSetSpec(logger, p, config, cg, shard, ruleConfigMapNames, tlsAssetSecrets)
+	p.SetCommonPrometheusFields(cpf)
```

Comment:
same question for me.

Reply:
We make a copy of CommonPrometheusFields when we call:
```go
cpf := p.GetCommonPrometheusFields()
```

Then we make changes to this copy, setting defaults if some fields aren't set:
```go
if cpf.PortName == "" {
	cpf.PortName = defaultPortName
}

if cpf.Replicas == nil {
	cpf.Replicas = &minReplicas
}
intZero := int32(0)
if cpf.Replicas != nil && *cpf.Replicas < 0 {
	cpf.Replicas = &intZero
}
```

Later we call `makeStatefulSetSpec`, but we don't use `cpf` as an argument, we use the same `p` object. Since `cpf` is a copy, the fields we just changed above aren't changed in the `p` object. So that's why `p.SetCommonPrometheusFields(cpf)` is needed 🙂

Reply:
Then couldn't we just do `cpf := &p.GetCommonPrometheusFields()`? Otherwise, if we want to follow this path I would at least advocate for a comment explaining the set

Reply:
Unfortunately we can't 😬 
![image](https://user-images.githubusercontent.com/24193764/219767820-ee834a08-ab33-4734-925a-9441c479339c.png)


Good point on adding a comment 👍

Reply:
Probably all the default values + validation should be handled at the OpenAPI spec level (like in https://github.com/prometheus-operator/prometheus-operator/pull/5350 for spec.portName). This way we can start with a clean state for the PrometheusAgent CRD and only set the defaults in the Prometheus controller. But like for the previous comment, this is fine for now.

Reply:
Yeah, that would make a lot more sense!